From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Sat, 11 Apr 2026 06:25:42 -0700
Subject: [PATCH] Paper: Add EntityLungeEvent

Original license: GPL-3.0-only
Original project: https://github.com/PaperMC/Paper
Paper commit: https://github.com/PaperMC/Paper/commit/67a54b3117aaa7e2cd81c6f777f5b1bccf4a98fc

Backported from Paper 26.1.1

Feature implementation for Paper#13634

diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 24df33ccd98bc9b1213f11574460db2f009578ca..e0124e1cba88868c73b2d08d21712df4bc79d645 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -361,7 +361,22 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
         applyEffects(
             this.getEffects(EnchantmentEffectComponents.POST_PIERCING_ATTACK),
             entityContext(level, enchantmentLevel, entity, entity.position()),
-            enchantmentEntityEffect -> enchantmentEntityEffect.apply(level, enchantmentLevel, item, entity, entity.position())
+            // Paper start - EntityLungeEvent
+            e -> {
+                int enchLevel = enchantmentLevel;
+                if (entity instanceof LivingEntity livingEntity) {
+                    final io.papermc.paper.event.entity.EntityLungeEvent event = new io.papermc.paper.event.entity.EntityLungeEvent(livingEntity.getBukkitLivingEntity(), enchLevel);
+
+                    if (!event.callEvent()) {
+                        return;
+                    }
+
+                    enchLevel = event.getLungePower();
+                }
+
+                e.apply(level, enchLevel, item, entity, entity.position());
+            }
+            // Paper end - EntityLungeEvent
         );
     }
 
